pbsmetrics/go_metrics.go

```diff
@@ -317,6 +317,10 @@ func (me *Metrics) RecordRequest(labels Labels) {
 
 func (me *Metrics) RecordImps(labels Labels, numImps int) {
 	me.ImpMeter.Mark(int64(numImps))
+	me.ImpsTypeBanner.Mark(int64(labels.BannerImps))
```

Comment:
Might be more efficient to put in if >0 on the counters before calling `Mark()` I am not sure that `Mark()` is smart enough to check for a no-op before engaging in all the thread safe concurrency machinery.

Reply:
ok, working on it

Reply:
Done

Reply:
Where is this done, and where is `RecordImps()` called from?

Reply:
I'm sorry Hans, those changes I had not commited. The latest version has been pushed. `deps.metricsEngine.RecordImps(labels, numImps)` gets called from every endpoint's auction function 
**endpoints/openrtb2/auction.go**
```
  75 func (deps *endpointDeps) Auction(w http.ResponseWriter, r *http.Request, _ httprouter.Params) {
  76
  77 +-- 10 lines: ao := analytics.AuctionObject{-----------------------------------------------------
  87     // to compute the auction timeout.
  88     start := time.Now()
  89     labels := pbsmetrics.Labels{
  90         Source:        pbsmetrics.DemandUnknown,
  91         RType:         pbsmetrics.ReqTypeORTB2Web,
  92         BannerImps:    0,
  93         VideoImps:     0,
  94         AudioImps:     0,
  95         NativeImps:    0,
  96         PubID:         "",
  97         Browser:       pbsmetrics.BrowserOther,
  98         CookieFlag:    pbsmetrics.CookieFlagUnknown,
  99         RequestStatus: pbsmetrics.RequestStatusOK,
 100     }
 101     numImps := 0
 102     defer func() {
 103         deps.metricsEngine.RecordRequest(labels)
 104         deps.metricsEngine.RecordImps(labels, numImps)  //<--
 105         deps.metricsEngine.RecordRequestTime(labels, time.Since(start))
 106         deps.analytics.LogAuctionObject(&ao)
 107     }()
 108
endpoints/openrtb2/auction.go
```
**endpoints/openrtb2/amp_auction.go**
```
 74 func (deps *endpointDeps) AmpAuction(w http.ResponseWriter, r *http.Request, _ httprouter.Params) {
 75
 76 +-- 14 lines: ao := analytics.AmpObject{----------------------------------------------------------
 90     start := time.Now()
 91     labels := pbsmetrics.Labels{
 92         Source:        pbsmetrics.DemandUnknown,
 93         RType:         pbsmetrics.ReqTypeAMP,
 94         BannerImps:    0,
 95         VideoImps:     0,
 96         AudioImps:     0,
 97         NativeImps:    0,
 98         PubID:         "",
 99         Browser:       pbsmetrics.BrowserOther,
100         CookieFlag:    pbsmetrics.CookieFlagUnknown,
101         RequestStatus: pbsmetrics.RequestStatusOK,
102     }
103     defer func() {
104         deps.metricsEngine.RecordRequest(labels)
105         deps.metricsEngine.RecordImps(labels, 1)  //<--
106         deps.metricsEngine.RecordRequestTime(labels, time.Since(start))
107         deps.analytics.LogAmpObject(&ao)
108     }()
endpoints/openrtb2/amp_auction.go
```
**endpoints/openrtb2/video_auction.go**
```
 66 func (deps *endpointDeps) VideoAuctionEndpoint(w http.ResponseWriter, r *http.Request, _ httprouter.Params) {
 67
 68     ao := analytics.AuctionObject{
 69         Status: http.StatusOK,
 70         Errors: make([]error, 0),
 71     }
 72
 73     start := time.Now()
 74     labels := pbsmetrics.Labels{
 75         Source:        pbsmetrics.DemandUnknown,
 76         RType:         pbsmetrics.ReqTypeVideo,
 77         BannerImps:    0,
 78         VideoImps:     0,
 79         AudioImps:     0,
 80         NativeImps:    0,
 81         PubID:         "",
 82         Browser:       pbsmetrics.BrowserOther,
 83         CookieFlag:    pbsmetrics.CookieFlagUnknown,
 84         RequestStatus: pbsmetrics.RequestStatusOK,
 85     }
 86     numImps := 0
 87     defer func() {
 88         deps.metricsEngine.RecordRequest(labels)
 89         deps.metricsEngine.RecordImps(labels, numImps)  //<--
 90         deps.metricsEngine.RecordRequestTime(labels, time.Since(start))
 91         deps.analytics.LogAuctionObject(&ao)
 92     }()
endpoints/openrtb2/video_auction.go
```